File: src/main/java/com/github/svegon/capi/event/render/NightVisionStrengthCallback.java
package com.github.svegon.capi.event.render;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.entity.LivingEntity;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;

public interface NightVisionStrengthCallback {
    Event<NightVisionStrengthCallback> EVENT = EventFactory.createArrayBacked(NightVisionStrengthCallback.class,
            (entity, tickDelta, callback) -> {}, listeners -> (entity, tickDelta, callback) -> {
        for (NightVisionStrengthCallback listener : listeners) {
            listener.getNightVisionStrength(entity, tickDelta, callback);

            if (callback.isCancelled()) {
                return;
            }
        }
    });

    void getNightVisionStrength(LivingEntity entity, float tickDelta, CallbackInfoReturnable<Float> callback);
}

File: src/main/kotlin/io/github/svegon/mclientapi/event/render/NightVisionStrengthCallback.kt
package io.github.svegon.capi.event.render

import io.github.svegon.capi.event.render.NightVisionStrengthCallback
import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.entity.LivingEntity
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable
import java.util.function.Function
import kotlin.Array
import kotlin.Float
import kotlin.invoke

interface NightVisionStrengthCallback {
    fun getNightVisionStrength(entity: LivingEntity?, tickDelta: Float, callback: CallbackInfoReturnable<Float?>?)

    companion object {
        val EVENT: Event<NightVisionStrengthCallback?> = EventFactory.createArrayBacked(
            NightVisionStrengthCallback::class.java,
            NightVisionStrengthCallback { entity: LivingEntity?, tickDelta: Float, callback: CallbackInfoReturnable<Float?>? -> },
            Function { listeners: Array<NightVisionStrengthCallback?>? ->
                NightVisionStrengthCallback { entity: LivingEntity?, tickDelta: Float, callback: CallbackInfoReturnable<Float?>? ->
                    for (listener in listeners) {
                        listener.getNightVisionStrength(entity, tickDelta, callback)

                        if (callback.isCancelled()) {
                            return@NightVisionStrengthCallback
                        }
                    }
                }
            })
    }
}
